src/nvim/diff.c

```diff
@@ -1104,21 +1185,32 @@ static int diff_file_internal(diffio_T *diffio)
 /// Make a diff between files "tmp_orig" and "tmp_new", results in "tmp_diff".
 ///
 /// @param dio
+/// @param idx_orig Buffer number of original file.
+/// @param idx_new Buffer number of new version of the file.
 ///
 /// @return OK or FAIL
-static int diff_file(diffio_T *dio)
+static int diff_file(diffio_T *dio, const int idx_orig, const int idx_new)
 {
   char *tmp_orig = (char *)dio->dio_orig.din_fname;
   char *tmp_new = (char *)dio->dio_new.din_fname;
   char *tmp_diff = (char *)dio->dio_diff.dout_fname;
+  list_T* chunks;
+
   if (*p_dex != NUL) {
-    // Use 'diffexpr' to generate the diff file.
-    eval_diff(tmp_orig, tmp_new, tmp_diff);
-    return OK;
+    // Evaluate function with `in_bufnr` and `new_bufnr` arguments.
+    if (diffexpr_internal_prefix()) {
+      // Ignore the first two '->' characters of the option,
+      // they are not part of the expression to evaluate.
+      chunks = eval_idifffn(idx_orig, idx_new, p_dex + 2);
+    } else {
+      // Otherwise use 'diffexpr' to generate the diff file.
+      eval_diff(tmp_orig, tmp_new, tmp_diff);
+      return OK;
+    }
   }
   // Use xdiff for generating the diff.
   if (dio->dio_internal) {
-    return diff_file_internal(dio);
+    return diff_file_internal(dio, chunks);
```

Comment:
This looks like it would break if `diffopt` doesn't have `internal` and `diffexpr=->...`?

You would solve this by returning straight after running `eval_idifffn`.

Reply:
Yupe, that's confusion on my side. I thought that `diff_file` was supposed to both fill up `dio` data then display the chunks, but it's only the first and the second is rather `diff_read`. So yes, I can return earlier.